### DISCUSSION TITLE
generating pandas table and visualising it in a report

### DISCUSSION THREAD

**Discussion by Justinas365** [quarto-dev/quarto-cli#1716] [Q&A]
Hi, is there a way to visualise nicely a pandas dataframe generated in python?

I want to create a function that does what is needed and returns me a pandas dataframe. The dataframe will involve some styling later on as well, such as highlighting cells automatically etc. (as per image below).
![image](https://user-images.githubusercontent.com/60322278/183291305-4eeae589-2afc-43a3-ac79-affb1ca2947e.png)

Searched the entire internet, but I dont think I managed to get what I needed, nor I am able to find what I want by testing different approaches. How would you make beautiful tables in Quarto using Python?

Thanks

**Comment by jjallaire** [comment]
You can use the `IPython.display` module to generate raw output (including HTML output). Documentation on this is here: https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html

We we will be working on some new Python packages to make this easier to do in pure Python (cc @rich-iannone) however these aren't yet available.

**Comment by Justinas365** [comment]
Appologies, didnt want to create a new topic. 

After long hours of researching, i discovered that:
 - You can make nice layout tables with extrakable in Rmarkdown
 - You can also define the way your report looks by using a custom css style, then print the report to PDF using pagedown.

Question: are the two options above planned for Quarto as well? Im keen to avoid LaTeX if possible, so that would be an excellent option

**Reply by rgaiacs** [reply]
Hi @Justinas365,

You can't define it in the YAML header but you can implement it yourself.

> You can make nice layout tables with extrakable in Rmarkdown

``````
```{r}
df |> extrakable::function()
```
``````

> You can also define the way your report looks by using a custom css style

In the YAML header,

```
format:
  html:
    css: custom-style.css
```

**Reply by Justinas365** [reply]
Thanks.
`format:
      html:
           css: custom-style.css`

The above is suitable for HTML only. Can I print the document into a PDF?

Thanks, Justinas

**Reply by jjallaire** [reply]
Pandas does indeed some methods for customizing PDF output but I'm not familiar with how they work.

**Reply by jjallaire** [reply]
We are also indeed planning on supporting paged.js in Quarto but not likely before the end of the year.

**Comment by kompre** [comment]
I would like too to use the pandas styler methods to style a dataframe with highlights, formats, etc. 

I've tried all day playing with pandas styler and it kinda works, because I get a formatted output that get rendered by quarto in the html and pdf, but it looks ugly in the html because it doesn't have the same table style of rest of the document (e.g. a markdown table).  Also if I have a math expression in the table enclosed in $...$ sign, quarto doesn't render it like a math expression, not even in the pdf version. But if I generate the html via a jupyter notebook, then the math expressions are correctly rendered.

I was able to generate a nice looking table using the builtin pandas' method `to_markdown()` that uses the `tabulate` package to generate a markdown  table, and then displaying such table with `Markdown()` from  `IPython.display`. The issue with this method is that the style method return a Styler object which has no function `to_markdown()`, which instead belong to a DataFrame object, so the styler can't be used; I can achieve some formatting passing some kwargs to the underlying `tabulate` via the `to_markdown()` method, but it seems to be more limited in scope.

**Reply by cscheid** [reply]
@kompre can you create an issue with the case that fails for you? We are going to be working on this in the next few months, so I'd love to have as much input from Python table users as we can. Thank you!

**Comment by dkgaraujo** [comment]
Just wondering if there is any news on that front (sorry if I missed it)

**Reply by mcanouil** [reply]
What do you mean exactly?

FYI, https://posit-dev.github.io/great-tables

**Reply by faysou** [reply]
@mcanouil impressive link
